### DISCUSSION TITLE
How to force the migrations to apply while developing Module

### DISCUSSION THREAD

**Discussion by vnetonline** [oqtane/oqtane.framework#2987] [General] [answered]
What is the best way to force migrations without creating multiple migration files while developing a module … atm I have to blow away the DB and start new install 

Surely there is a better way to force migration ??

@leigh-pointer how do you get around this ?

**Comment by leigh-pointer** [comment]
Personally i do it all manually! S bit irritating but i havent looked inti doing it differently.

**Reply by vnetonline** [reply]
When you say you do it manually you created new columns in db manually ??

**Reply by leigh-pointer** [reply]
@vnetonline yes

**Reply by vnetonline** [reply]
Hmmmm tedious what does ActivateModule do ?? Does it run migration ??

Need to find a way of running migrations manually developing experience is very tedious

**Reply by leigh-pointer** [reply]
Activate module will add an instance to the page where the Create Module was! Create Module is also a place holder.

**Reply by vnetonline** [reply]
What triggers the migration ??

**Reply by leigh-pointer** [reply]
For a module it will be the version number

**Comment by sbwalker** [comment, marked as answer]
Migrations are based on version number. Once a migration has been executed, it will be recorded in the __EFMigrationsHistory table in the database (this is standard EF Core behavior). The __EFMigrationsHistory table allows EF Core to know if a migration has already been previously executed. In addition, Oqtane uses the Version identified in the ModuleInfo class in conjunction with the Version stored in the ModuleDefinition table to determine if a module needs to be upgraded or not. 

During development there are variety of ways to deal with migrations:

- include them in a migration class but implement them manually in the database development (like @leigh-pointer mentioned). When the project nears an official release you obviously need to test the actual migration class on a clean database to ensure it behaves as expected.
- create a migration class for each database change - with a unique version. Ensure you also change the version in ModuleInfo to match the latest migration version. When the framework starts up it will execute the migration. Later you can collapse all of the migrations into a single class file if you want to.
- manually modify database Version number in ModuleDefinition and delete record from __EFMigrationsHistory to re-execute a migration. This is often problematic as migrations are generally not re-executable.
